File: sdks/js/src/core/sanitize.ts
import { createHash } from 'node:crypto';
import type { Event } from './event.ts';

/**
 * Sanitize an event for safe schema encoding.
 * Clones the event, applies sensitive→[REDACTED], hashValue→sha256, drop→delete.
 * The original event is never mutated.
 */
export function sanitizeEvent(event: Event): Event {
  const clone = event.clone();

  // Apply sensitive keys → redact values
  for (const key of clone.sensitiveKeys) {
    if (key in clone.attrs) {
      clone.attrs[key] = '[REDACTED]';
    }
  }

  // Apply hash keys → sha256 hash values
  for (const key of clone.hashKeys) {
    if (key in clone.attrs) {
      const val = clone.attrs[key];
      if (typeof val === 'string') {
        clone.attrs[key] = createHash('sha256').update(val).digest('hex');
      }
    }
  }

  // Remove dropped keys
  for (const key of clone.droppedKeys) {
    delete clone.attrs[key];
  }

  return clone;
}

File: sdks/go/src/core/sanitize.go
package core

// SanitizeEvent clones the event and applies the global config's redactor
// and security settings. The original event is never mutated.
func SanitizeEvent(ev *Event) *Event {
	if ev == nil {
		return nil
	}
	clone := ev.Clone()
	if clone == nil {
		return nil
	}
	clone.logger = nil
	cfg := Default().cfg
	if cfg.Redactor != nil {
		clone.Attrs = applyRedactor(clone.Attrs, cfg.Redactor)
	}
	if cfg.Security.MaxAttrCount > 0 || cfg.Security.MaxFieldBytes > 0 || (cfg.Security.RedactByDefault && !cfg.Security.AllowPII) {
		clone.Attrs = applySecurity(clone.Attrs, cfg.Security)
	}
	return clone
}
